### DISCUSSION TITLE
User should receive an e-mail with a link to confirm the submission

### DISCUSSION THREAD

**Discussion by whoiscado** [solspace/craft-freeform#222] [Questions & Answers] [answered]
Hey!

I ran into a problem and can't find a solution for it...

**target state:**
- user fills out form
  - submission is "unactive"
- user receives a mail `"Thanks for your submission. Please click the link below to confirm your submission"`
   - user clicks the link
   - submission and the fieldset in the form turns "active"
  

Does anyone know how I can achieve that target state and how I can set up the link in the mail?

Thanks a lot for any kind of help!

**Comment by kjmartens** [comment, marked as answer]
Hi @whoiscado,

What you would have to do is initially set the form submissions by users to be something like `pending`. Then, in an email notification template, be sure to include a URL to another template that loads the form and include `{{ token }}` at the end of the URL (which will parse as the submission token for editing).

Then in your template (specific for "confirming" (editing) the submission), be sure to grab the token from the URL and specify the `status` parameter in the form. So something like this:

```
{% set submissionToken = craft.app.request.segment(3) %}

{% set form = craft.freeform.form("myForm", {
    returnUrl: "{{ siteUrl }}wherever/success",
    submissionToken: submissionToken|default(null),
    status: "open"
}) %}
```

It won't be the exact same as clicking a single link to "confirm" it, as this would just present the same form but in edit mode, so at best, you could hide the rest of the fields and just leave behind a "submit" button for the user (which would require that extra click).

And finally, since you're dealing with email notifications, you will need to suppress certain ones at certain times, so be sure to see this (otherwise the user "confirming" (editing) their submission will trigger another email telling them to confirm the same submission again forever):
https://docs.solspace.com/craft/freeform/v3/overview/submissions.html#suppression-of-notifications-and-more

The other option would be to build a custom module using Freeform's [developer events](https://docs.solspace.com/craft/freeform/v3/developer/events.html) to handle this more robustly. 🙂

**Reply by whoiscado** [reply]
Hi @kjmartens,

thanks for your answer! I'll try it in the next days and mark it as solved here. :)
Weird that Freeform has no official double opt-in options as f.e. for newsletters. 

Have a great day!

**Reply by whoiscado** [reply]
Hey @kjmartens

Is it possible to edit the submission in the twig template?

So user gets a mail after submitting the submission.
The submission is `pending`.
User clicks on the `link.com/example?{{ token }}`.
Gets shown a message "Thanks for confirming your submission" and doesn't need to click anything else on the site.
The submission is `open`.

Sadly I didn't find any way to do it directly in the .twig template.

**Reply by kjmartens** [reply]
There sure is! Check out the docs for Submission Editing here:
https://docs.solspace.com/craft/freeform/v3/overview/submission-editing.html

**Reply by whoiscado** [reply]
Hey @kjmartens,

thanks for your quick reply and help! 
Sadly I'm unfamiliar with twig and I'm unsure how to achieve, that the submission's status get's automatically changed to `open` after the user clicked on the confirmation link in the mail and gets redirected to my custom `/confirmation` route with the corresponding template.

I'm trying to achieve it with something like this:
```twig
{% set submissionToken = craft.app.request.getQueryParam("token") %}
{% set submissions = craft.freeform.submissions({
    status: ["pending"],
}) %}
{% for submission in submissions %}
        {% if submission.token is same as (submissionToken) %}
            {% do craft.freeform.setSubmissionStatusAction.performAction(submission.status, open) %}
        {% endif %}
{% endfor %}
```
Could you please tell me which function is the correct one that saves the submission status to open?

Thanks a lot for your help and time!

**Reply by kjmartens** [reply]
Sorry, I forgot that your request was to get this to happen in a single click / page load. This isn't currently possible with Freeform using Twig. In order to have something like this work, you'd need to build a custom module that would handle this.

**Reply by whoiscado** [reply]
Thanks @kjmartens. That's what I thought, too.

Do you happen to know which freeform service gets used (for example when clicking manually in the submission in the Craft CP) for changing the status and saves it to the database?

**Reply by kjmartens** [reply]
This would be a starting point:
https://docs.solspace.com/craft/freeform/v3/developer/freeform-api.html#getting-freeform-forms-and-submissions-inside-your-controller

We also offer custom development services as well. We could potentially custom-build this module for you as well. If you're interested in that, please send us an email at support@solspace.com. 🙂
